doc/specifications/proposed/replicator/architecture.rst

```diff
@@ -0,0 +1,142 @@
+==========
+Replicator
+==========
+
+This document describes a common synchronization mechanism between two systems,
+which can cope with write failures, network downtimes and other types of
+partitions. The approach only works with a single write master (Single Source
+of Truth) and any number of read slaves.
+
+In our case we want to use this approach to synchronize the content database
+with a search index.
+
+General Workings
+================
+
+The general scheme for synchronizations is fairly simple:
+
+.. image:: synchronize.png
+
+It generally can be described as a three step process:
+
+1) Fetch last revision from target
+
+   The target is asked for the last revision the target knows. This is the
+   revision of the last successful write operation. For the first request the
+   target may return something like ``null``.
+
+2) Fetch all changes since that last revision from source
+
+   The source (write-master) is asked for all updates since the last revision.
+   A limit, to limit the transfer size, may also be applied. The results
+   **must** be sorted by their revision.
+
+3) Commit all updates on target
+
+   All fetched updates are committed on the target. The target **must not**
+   store a revision if the corresponding update did not succeed.
+
+   This is usually possible to implement using transactions. Another way is to
+   store the revision directly in the received update (document). This way, if
+   the document update is lost, the revision will also be "lost".
+
+Revisions
+---------
+
+Revisions **must** always be **globally strictly monotonically** increasing.
+
+If you have only one write-master time (including milliseconds) might work,
+unless there are multiple commits in the same millisecond or some NTP daemon
+stops time or makes it go backwards. Time will never be synchronous across
+multiple systems, so **do not** rely on that.
+
+A simple auto-increment in the write-master might be the simplest solution. You
+should take care for possible overflows, though. The revision might increase
+quite fast in scenarios with a lot of updates.
+
+Idempotent Updates
+------------------
+
+In some cases it is helpful if all updates are idempotent, thus applying the
```

Comment:
Tip: ALL updates MUST be idempotent. This makes the protocol much more stable, reliable, easy to implement.

Ideally you would have a log somewhere of the unexpected situations (eg. master has deleted a content but that content is not there on the slave), and design your apis so that those warnings could easily be turned into blockers if needed.